src/components/UserContext/UserContext.tsx

```diff
@@ -0,0 +1,137 @@
+import React, { useState, useCallback } from 'react';
+import { getTodos, getUser } from '../../api';
+import { SortTypes } from '../../types/SortTypes';
+import { Todo } from '../../types/Todo';
+import { User } from '../../types/User';
+import { debounce } from '../utils/debounce';
+
+type Context = {
+  todos: Todo[],
+  selectedUser: number,
+  selectedTodoId: number,
+  userInfo: User | null,
+  sortType: SortTypes,
+  query: string,
+  appliedQuery: string,
+  isLoaded: boolean,
+  isError: boolean,
+
+  getUserFromServer: () => void,
+  handleClose: () => void,
+  getTodosFromServer: () => void,
+  changeSortType: (event: React.ChangeEvent<HTMLSelectElement>) => void,
+  applyQuery: ((arg: string) => void),
+  handleSearch: (event: React.ChangeEvent<HTMLInputElement>) => void,
+  clearSearchBar: () => void,
+  selectCurrentTodo: (id: number, userId: number) => void
+};
+
+export const UserContext = React.createContext<Context>({
+  todos: [],
+  selectedUser: 0,
+  selectedTodoId: 0,
+  userInfo: null,
+  sortType: SortTypes.ALL,
+  query: '',
+  appliedQuery: '',
+  isLoaded: true,
+  isError: false,
+
+  getUserFromServer: () => {},
+  handleClose: () => {},
+  getTodosFromServer: () => {},
+  changeSortType: () => {},
+  applyQuery: () => {},
+  handleSearch: () => {},
+  clearSearchBar: () => {},
+  selectCurrentTodo: () => {},
+});
+
+export const UserProvider: React.FC = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [selectedUser, setSelectedUser] = useState(0);
+  const [selectedTodoId, setSelectedTodoId] = useState(0);
+  const [userInfo, setUserInfo] = useState<User | null>(null);
+  const [sortType, setSortType] = useState<SortTypes>(SortTypes.ALL);
+  const [query, setQuery] = useState('');
+  const [appliedQuery, setAppliedQuery] = useState('');
```

Comment:
variables that can be changeable a lot of times will be better to move to the component, where you handle them
in context you should store not all variables for your app, just the main one, the main data of your app (like todo and user for this one)

other variables, related to the query and to the info, was data loaded or no, you can move to the state of the needed component

Reply:
In my case I throw query and applied query from TodoFilter to TodoList and they connect only through Context, will try to connect them in other way tomorrow, now just separate User and Todo in two Contexts